File: backend/api/audit.py
"""
Audit API endpoints for SoftBankCashWire
"""
from flask import Blueprint, request, jsonify, g
from datetime import datetime, timedelta
from services.audit_service import AuditService
from middleware.auth_middleware import auth_required, finance_required, admin_required
from models import db

audit_bp = Blueprint('audit', __name__)

@audit_bp.route('/logs', methods=['GET'])
@finance_required
@auth_required
def get_audit_logs():
    """
    Get audit logs with filtering and pagination (Finance team only)
    
    Query Parameters:
        - user_id: Filter by user ID
        - action_type: Filter by action type
        - entity_type: Filter by entity type
        - start_date: Start date (ISO format)
        - end_date: End date (ISO format)
        - ip_address: Filter by IP address
        - severity: Filter by severity (INFO, WARNING, ERROR, CRITICAL)
        - page: Page number (default: 1)
        - per_page: Items per page (default: 50, max: 1000)
        - include_system_events: Include system events (default: true)
    
    Returns:
        JSON with audit logs and pagination info
    """
    try:
        # Parse query parameters
        filters = {}
        
        if request.args.get('user_id'):
            filters['user_id'] = request.args.get('user_id')
        
        if request.args.get('action_type'):
            filters['action_type'] = request.args.get('action_type')
        
        if request.args.get('entity_type'):
            filters['entity_type'] = request.args.get('entity_type')
        
        if request.args.get('ip_address'):
            filters['ip_address'] = request.args.get('ip_address')
        
        if request.args.get('severity'):
            filters['severity'] = request.args.get('severity')
        
        # Date filters
        if request.args.get('start_date'):
            try:
                filters['start_date'] = datetime.fromisoformat(
                    request.args.get('start_date').replace('Z', '+00:00')
                )
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_DATE_FORMAT',
                        'message': 'start_date must be in ISO format'
                    }
                }), 400
        
        if request.args.get('end_date'):
            try:
                filters['end_date'] = datetime.fromisoformat(
                    request.args.get('end_date').replace('Z', '+00:00')
                )
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_DATE_FORMAT',
                        'message': 'end_date must be in ISO format'
                    }
                }), 400
        
        # Pagination
        if request.args.get('page'):
            try:
                filters['page'] = int(request.args.get('page'))
                if filters['page'] < 1:
                    raise ValueError()
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_PAGE',
                        'message': 'Page must be a positive integer'
                    }
                }), 400
        
        if request.args.get('per_page'):
            try:
                filters['per_page'] = int(request.args.get('per_page'))
                if filters['per_page'] < 1 or filters['per_page'] > 1000:
                    raise ValueError()
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_PER_PAGE',
                        'message': 'per_page must be between 1 and 1000'
                    }
                }), 400
        
        # Include system events
        filters['include_system_events'] = request.args.get('include_system_events', 'true').lower() == 'true'
        
        # Get audit logs
        result = AuditService.get_audit_logs(filters)
        
        return jsonify(result), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'AUDIT_LOGS_ERROR',
                'message': f'Failed to get audit logs: {str(e)}'
            }
        }), 500

@audit_bp.route('/reports/generate', methods=['POST'])
@finance_required
@auth_required
def generate_audit_report():
    """
    Generate comprehensive audit report (Finance team only)
    
    Expected JSON:
        {
            "start_date": "2024-01-01T00:00:00Z",
            "end_date": "2024-01-31T23:59:59Z",
            "report_type": "COMPREHENSIVE"  // COMPREHENSIVE, TRANSACTIONS, SECURITY, USER_ACTIVITY
        }
    
    Returns:
        JSON with audit report data
    """
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({
                'error': {
                    'code': 'MISSING_DATA',
                    'message': 'Request body is required'
                }
            }), 400
        
        # Parse dates
        try:
            start_date = datetime.fromisoformat(data['start_date'].replace('Z', '+00:00'))
            end_date = datetime.fromisoformat(data['end_date'].replace('Z', '+00:00'))
        except (KeyError, ValueError):
            return jsonify({
                'error': {
                    'code': 'INVALID_DATES',
                    'message': 'start_date and end_date are required in ISO format'
                }
            }), 400
        
        # Validate date range
        if start_date >= end_date:
            return jsonify({
                'error': {
                    'code': 'INVALID_DATE_RANGE',
                    'message': 'start_date must be before end_date'
                }
            }), 400
        
        # Check if date range is too large (max 1 year)
        if (end_date - start_date).days > 365:
            return jsonify({
                'error': {
                    'code': 'DATE_RANGE_TOO_LARGE',
                    'message': 'Date range cannot exceed 365 days'
                }
            }), 400
        
        # Parse report type
        report_type = data.get('report_type', 'COMPREHENSIVE')
        valid_types = ['COMPREHENSIVE', 'TRANSACTIONS', 'SECURITY', 'USER_ACTIVITY']
        
        if report_type not in valid_types:
            return jsonify({
                'error': {
                    'code': 'INVALID_REPORT_TYPE',
                    'message': f'report_type must be one of: {", ".join(valid_types)}'
                }
            }), 400
        
        # Generate report
        report = AuditService.generate_audit_report(start_date, end_date, report_type)
        
        # Log report generation
        AuditService.log_user_action(
            user_id=g.current_user_id,
            action_type='AUDIT_REPORT_GENERATED',
            entity_type='AuditReport',
            new_values={
                'report_type': report_type,
                'start_date': start_date.isoformat(),
                'end_date': end_date.isoformat(),
                'generated_by': g.current_user.name if g.current_user else 'Unknown'
            }
        )
        
        return jsonify(report), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'REPORT_GENERATION_ERROR',
                'message': f'Failed to generate audit report: {str(e)}'
            }
        }), 500

@audit_bp.route('/statistics', methods=['GET'])
@finance_required
@auth_required
def get_audit_statistics():
    """
    Get audit statistics (Finance team only)
    
    Query Parameters:
        - days: Number of days to analyze (default: 30, max: 365)
    
    Returns:
        JSON with audit statistics
    """
    try:
        # Parse days parameter
        days = 30
        if request.args.get('days'):
            try:
                days = int(request.args.get('days'))
                if days < 1 or days > 365:
                    raise ValueError()
            except ValueError:
                return jsonify({
                    'error': {
                        'code': 'INVALID_DAYS',
                        'message': 'Days must be between 1 and 365'
                    }
                }), 400
        
        # Get statistics
        statistics = AuditService.get_audit_statistics(days)
        
        return jsonify(statistics), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'STATISTICS_ERROR',
                'message': f'Failed to get audit statistics: {str(e)}'
            }
        }), 500

@audit_bp.route('/integrity/verify', methods=['POST'])
@admin_required
@auth_required
def verify_audit_integrity():
    """
    Verify audit log integrity (Admin only)
    
    Returns:
        JSON with integrity check results
    """
    try:
        # Perform integrity check
        integrity_results = AuditService.verify_audit_integrity()
        
        # Log integrity check
        AuditService.log_user_action(
            user_id=g.current_user_id,
            action_type='AUDIT_INTEGRITY_CHECK',
            entity_type='AuditSystem',
            new_values={
                'total_logs_checked': integrity_results['total_logs_checked'],
                'issues_found': len(integrity_results['integrity_issues']),
                'overall_status': integrity_results['overall_status'],
                'checked_by': g.current_user.name if g.current_user else 'Unknown'
            }
        )
        
        return jsonify(integrity_results), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'INTEGRITY_CHECK_ERROR',
                'message': f'Failed to verify audit integrity: {str(e)}'
            }
        }), 500

@audit_bp.route('/cleanup', methods=['POST'])
@admin_required
@auth_required
def cleanup_old_logs():
    """
    Clean up old audit logs based on retention policy (Admin only)
    
    Expected JSON:
        {
            "retention_days": 2555  // Optional, default: 2555 (7 years)
        }
    
    Returns:
        JSON with cleanup results
    """
    try:
        data = request.get_json() or {}
        
        # Parse retention days
        retention_days = data.get('retention_days', 2555)  # Default: 7 years
        
        try:
            retention_days = int(retention_days)
            if retention_days < 1:
                raise ValueError()
        except (ValueError, TypeError):
            return jsonify({
                'error': {
                    'code': 'INVALID_RETENTION_DAYS',
                    'message': 'retention_days must be a positive integer'
                }
            }), 400
        
        # Minimum retention period check (e.g., 90 days)
        if retention_days < 90:
            return jsonify({
                'error': {
                    'code': 'RETENTION_TOO_SHORT',
                    'message': 'Retention period cannot be less than 90 days for compliance'
                }
            }), 400
        
        # Perform cleanup
        cleanup_result = AuditService.cleanup_old_audit_logs(retention_days)
        
        # Log cleanup operation
        AuditService.log_user_action(
            user_id=g.current_user_id,
            action_type='AUDIT_CLEANUP_PERFORMED',
            entity_type='AuditSystem',
            new_values={
                'retention_days': retention_days,
                'deleted_count': cleanup_result['deleted_count'],
                'success': cleanup_result['success'],
                'performed_by': g.current_user.name if g.current_user else 'Unknown'
            }
        )
        
        return jsonify(cleanup_result), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'CLEANUP_ERROR',
                'message': f'Failed to cleanup audit logs: {str(e)}'
            }
        }), 500

@audit_bp.route('/action-types', methods=['GET'])
@finance_required
@auth_required
def get_action_types():
    """
    Get available audit action types (Finance team only)
    
    Returns:
        JSON with categorized action types
    """
    try:
        action_types = AuditService.ACTION_TYPES
        
        return jsonify({
            'action_types': action_types,
            'total_categories': len(action_types),
            'total_action_types': sum(len(actions) for actions in action_types.values())
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'ACTION_TYPES_ERROR',
                'message': f'Failed to get action types: {str(e)}'
            }
        }), 500

@audit_bp.route('/export', methods=['POST'])
@finance_required
@auth_required
def export_audit_logs():
    """
    Export audit logs in various formats (Finance team only)
    
    Expected JSON:
        {
            "start_date": "2024-01-01T00:00:00Z",
            "end_date": "2024-01-31T23:59:59Z",
            "format": "CSV",  // CSV, JSON
            "filters": {  // Optional additional filters
                "user_id": "user-123",
                "action_type": "TRANSACTION_CREATED"
            }
        }
    
    Returns:
        JSON with export data or download link
    """
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({
                'error': {
                    'code': 'MISSING_DATA',
                    'message': 'Request body is required'
                }
            }), 400
        
        # Parse dates
        try:
            start_date = datetime.fromisoformat(data['start_date'].replace('Z', '+00:00'))
            end_date = datetime.fromisoformat(data['end_date'].replace('Z', '+00:00'))
        except (KeyError, ValueError):
            return jsonify({
                'error': {
                    'code': 'INVALID_DATES',
                    'message': 'start_date and end_date are required in ISO format'
                }
            }), 400
        
        # Parse format
        export_format = data.get('format', 'JSON').upper()
        if export_format not in ['CSV', 'JSON']:
            return jsonify({
                'error': {
                    'code': 'INVALID_FORMAT',
                    'message': 'Format must be CSV or JSON'
                }
            }), 400
        
        # Get filters
        filters = data.get('filters', {})
        filters.update({
            'start_date': start_date,
            'end_date': end_date,
            'per_page': 10000  # Large limit for export
        })
        
        # Get audit logs
        result = AuditService.get_audit_logs(filters)
        
        # Log export operation
        AuditService.log_user_action(
            user_id=g.current_user_id,
            action_type='AUDIT_LOGS_EXPORTED',
            entity_type='AuditSystem',
            new_values={
                'format': export_format,
                'start_date': start_date.isoformat(),
                'end_date': end_date.isoformat(),
                'record_count': len(result['audit_logs']),
                'exported_by': g.current_user.name if g.current_user else 'Unknown'
            }
        )
        
        if export_format == 'JSON':
            return jsonify({
                'format': 'JSON',
                'data': result['audit_logs'],
                'metadata': {
                    'export_timestamp': datetime.utcnow().isoformat(),
                    'record_count': len(result['audit_logs']),
                    'date_range': {
                        'start': start_date.isoformat(),
                        'end': end_date.isoformat()
                    }
                }
            }), 200
        
        elif export_format == 'CSV':
            # For CSV, return structured data that frontend can convert
            csv_data = []
            for log in result['audit_logs']:
                csv_row = {
                    'timestamp': log['created_at'],
                    'user_id': log.get('user_id', ''),
                    'user_name': log.get('user_name', ''),
                    'action_type': log['action_type'],
                    'entity_type': log['entity_type'],
                    'entity_id': log.get('entity_id', ''),
                    'ip_address': log.get('ip_address', ''),
                    'user_agent': log.get('user_agent', ''),
                    'severity': log.get('severity', 'INFO')
                }
                csv_data.append(csv_row)
            
            return jsonify({
                'format': 'CSV',
                'data': csv_data,
                'headers': list(csv_data[0].keys()) if csv_data else [],
                'metadata': {
                    'export_timestamp': datetime.utcnow().isoformat(),
                    'record_count': len(csv_data),
                    'date_range': {
                        'start': start_date.isoformat(),
                        'end': end_date.isoformat()
                    }
                }
            }), 200
        
    except Exception as e:
        return jsonify({
            'error': {
                'code': 'EXPORT_ERROR',
                'message': f'Failed to export audit logs: {str(e)}'
            }
        }), 500

# Error handlers for audit blueprint
@audit_bp.errorhandler(400)
def bad_request(error):
    """Handle bad request errors"""
    return jsonify({
        'error': {
            'code': 'BAD_REQUEST',
            'message': 'Invalid request format'
        }
    }), 400

@audit_bp.errorhandler(401)
def unauthorized(error):
    """Handle unauthorized errors"""
    return jsonify({
        'error': {
            'code': 'UNAUTHORIZED',
            'message': 'Authentication required'
        }
    }), 401

@audit_bp.errorhandler(403)
def forbidden(error):
    """Handle forbidden errors"""
    return jsonify({
        'error': {
            'code': 'FORBIDDEN',
            'message': 'Finance team access required'
        }
    }), 403

@audit_bp.errorhandler(500)
def internal_error(error):
    """Handle internal server errors"""
    db.session.rollback()
    return jsonify({
        'error': {
            'code': 'INTERNAL_ERROR',
            'message': 'Internal server error'
        }
    }), 500

File: frontend/src/services/audit.ts
/**
 * Audit service for SoftBankCashWire frontend
 */
import apiClient, { handleApiError } from './api'

export interface AuditLog {
  id: string
  user_id?: string
  user_name?: string
  action_type: string
  entity_type: string
  entity_id?: string
  old_values?: Record<string, any>
  new_values?: Record<string, any>
  ip_address?: string
  user_agent?: string
  created_at: string
  is_system_event: boolean
  severity: 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL'
  changes?: Record<string, { old: any; new: any }>
}

export interface AuditLogFilters {
  user_id?: string
  action_type?: string
  entity_type?: string
  start_date?: string
  end_date?: string
  ip_address?: string
  severity?: 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL'
  page?: number
  per_page?: number
  include_system_events?: boolean
}

export interface AuditLogList {
  audit_logs: AuditLog[]
  pagination: {
    page: number
    per_page: number
    total: number
    pages: number
    has_prev: boolean
    has_next: boolean
    prev_num?: number
    next_num?: number
  }
}

export interface AuditReport {
  report_type: string
  period: {
    start_date: string
    end_date: string
    duration_days: number
  }
  generated_at: string
  generated_by: string
  user_activity?: {
    total_users_active: number
    total_user_actions: number
    user_details: Record<string, {
      user_name: string
      user_email: string
      total_actions: number
      action_types: Record<string, number>
      login_count: number
      transaction_count: number
      last_activity?: string
    }>
  }
  transactions?: {
    total_transaction_events: number
    transaction_types: Record<string, number>
    failed_transactions: number
    bulk_transfers: number
    total_amount_processed: number
  }
  security?: {
    total_security_events: number
    failed_logins: number
    successful_logins: number
    security_alerts: number
    unique_ip_addresses: number
    suspicious_activities: Array<{
      timestamp: string
      action_type: string
      user_id?: string
      ip_address?: string
      details?: Record<string, any>
    }>
  }
  system?: {
    total_system_events: number
    system_errors: number
    system_warnings: number
    system_info: number
    event_types: Record<string, number>
  }
  compliance?: {
    total_audit_entries: number
    data_integrity_checks: {
      entries_with_timestamps: number
      entries_with_user_context: number
      entries_with_ip_tracking: number
    }
    retention_compliance: {
      oldest_entry?: string
      newest_entry?: string
      retention_period_days?: number
    }
    audit_coverage: {
      user_actions_logged: number
      system_events_logged: number
      transaction_events_logged: number
    }
  }
}

export interface AuditStatistics {
  period_days: number
  total_entries: number
  user_actions: number
  system_events: number
  action_type_breakdown: Record<string, number>
  entity_type_breakdown: Record<string, number>
  user_activity: Record<string, {
    user_name: string
    action_count: number
  }>
  daily_activity: Record<string, number>
  security_events: number
  transaction_events: number
}

export interface IntegrityCheckResult {
  total_logs_checked: number
  integrity_issues: Array<{
    log_id: string
    issue: string
    severity: 'LOW' | 'MEDIUM' | 'HIGH'
  }>
  missing_timestamps: number
  missing_action_types: number
  orphaned_user_references: number
  data_consistency_issues: number
  overall_status: 'HEALTHY' | 'WARNING' | 'CRITICAL'
}

export interface CleanupResult {
  success: boolean
  deleted_count: number
  cutoff_date?: string
  message: string
  error?: string
}

export interface ActionTypes {
  action_types: {
    USER_ACTIONS: string[]
    TRANSACTION_ACTIONS: string[]
    ACCOUNT_ACTIONS: string[]
    EVENT_ACTIONS: string[]
    MONEY_REQUEST_ACTIONS: string[]
    SYSTEM_ACTIONS: string[]
  }
  total_categories: number
  total_action_types: number
}

export interface ExportRequest {
  start_date: string
  end_date: string
  format: 'CSV' | 'JSON'
  filters?: Record<string, any>
}

export interface ExportResult {
  format: 'CSV' | 'JSON'
  data: any[]
  headers?: string[]
  metadata: {
    export_timestamp: string
    record_count: number
    date_range: {
      start: string
      end: string
    }
  }
}

class AuditService {
  /**
   * Get audit logs with filtering and pagination
   */
  static async getAuditLogs(filters?: AuditLogFilters): Promise<AuditLogList> {
    try {
      const params = new URLSearchParams()
      
      if (filters) {
        Object.entries(filters).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            params.append(key, value.toString())
          }
        })
      }

      const response = await apiClient.get(`/audit/logs?${params.toString()}`)
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Generate comprehensive audit report
   */
  static async generateAuditReport(
    startDate: string,
    endDate: string,
    reportType: 'COMPREHENSIVE' | 'TRANSACTIONS' | 'SECURITY' | 'USER_ACTIVITY' = 'COMPREHENSIVE'
  ): Promise<AuditReport> {
    try {
      const response = await apiClient.post('/audit/reports/generate', {
        start_date: startDate,
        end_date: endDate,
        report_type: reportType
      })
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Get audit statistics
   */
  static async getAuditStatistics(days: number = 30): Promise<AuditStatistics> {
    try {
      const response = await apiClient.get(`/audit/statistics?days=${days}`)
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Verify audit log integrity
   */
  static async verifyAuditIntegrity(): Promise<IntegrityCheckResult> {
    try {
      const response = await apiClient.post('/audit/integrity/verify')
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Clean up old audit logs
   */
  static async cleanupOldLogs(retentionDays: number = 2555): Promise<CleanupResult> {
    try {
      const response = await apiClient.post('/audit/cleanup', {
        retention_days: retentionDays
      })
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Get available action types
   */
  static async getActionTypes(): Promise<ActionTypes> {
    try {
      const response = await apiClient.get('/audit/action-types')
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Export audit logs
   */
  static async exportAuditLogs(exportRequest: ExportRequest): Promise<ExportResult> {
    try {
      const response = await apiClient.post('/audit/export', exportRequest)
      return response.data
    } catch (error) {
      throw handleApiError(error)
    }
  }

  /**
   * Get severity color for UI
   */
  static getSeverityColor(severity: string): 'success' | 'warning' | 'error' | 'info' {
    switch (severity.toUpperCase()) {
      case 'CRITICAL':
      case 'ERROR':
        return 'error'
      case 'WARNING':
        return 'warning'
      case 'INFO':
        return 'info'
      default:
        return 'success'
    }
  }

  /**
   * Format action type for display
   */
  static formatActionType(actionType: string): string {
    return actionType
      .split('_')
      .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
      .join(' ')
  }

  /**
   * Get entity type icon
   */
  static getEntityTypeIcon(entityType: string): string {
    switch (entityType.toLowerCase()) {
      case 'user':
        return '👤'
      case 'account':
        return '💰'
      case 'transaction':
        return '💸'
      case 'eventaccount':
        return '🎉'
      case 'moneyrequest':
        return '💳'
      case 'system':
        return '⚙️'
      case 'security':
        return '🔒'
      default:
        return '📄'
    }
  }

  /**
   * Check if audit log is a security event
   */
  static isSecurityEvent(auditLog: AuditLog): boolean {
    return (
      auditLog.action_type.includes('SECURITY') ||
      auditLog.action_type.includes('LOGIN') ||
      auditLog.entity_type === 'Security' ||
      auditLog.severity === 'CRITICAL'
    )
  }

  /**
   * Check if audit log is a transaction event
   */
  static isTransactionEvent(auditLog: AuditLog): boolean {
    return (
      auditLog.action_type.includes('TRANSACTION') ||
      auditLog.entity_type === 'Transaction'
    )
  }

  /**
   * Format timestamp for display
   */
  static formatTimestamp(timestamp: string): string {
    return new Date(timestamp).toLocaleString()
  }

  /**
   * Get time ago string
   */
  static getTimeAgo(timestamp: string): string {
    const now = new Date()
    const time = new Date(timestamp)
    const diffInSeconds = Math.floor((now.getTime() - time.getTime()) / 1000)

    if (diffInSeconds < 60) {
      return 'Just now'
    } else if (diffInSeconds < 3600) {
      const minutes = Math.floor(diffInSeconds / 60)
      return `${minutes} minute${minutes > 1 ? 's' : ''} ago`
    } else if (diffInSeconds < 86400) {
      const hours = Math.floor(diffInSeconds / 3600)
      return `${hours} hour${hours > 1 ? 's' : ''} ago`
    } else {
      const days = Math.floor(diffInSeconds / 86400)
      return `${days} day${days > 1 ? 's' : ''} ago`
    }
  }

  /**
   * Filter audit logs by text search
   */
  static searchAuditLogs(auditLogs: AuditLog[], searchText: string): AuditLog[] {
    if (!searchText.trim()) {
      return auditLogs
    }

    const searchLower = searchText.toLowerCase()
    
    return auditLogs.filter(log => {
      return (
        log.action_type.toLowerCase().includes(searchLower) ||
        log.entity_type.toLowerCase().includes(searchLower) ||
        log.user_name?.toLowerCase().includes(searchLower) ||
        log.ip_address?.includes(searchText) ||
        log.entity_id?.includes(searchText) ||
        JSON.stringify(log.new_values || {}).toLowerCase().includes(searchLower)
      )
    })
  }

  /**
   * Sort audit logs by various criteria
   */
  static sortAuditLogs(
    auditLogs: AuditLog[], 
    sortBy: 'timestamp' | 'user' | 'action' | 'severity',
    sortOrder: 'asc' | 'desc' = 'desc'
  ): AuditLog[] {
    const sorted = [...auditLogs].sort((a, b) => {
      let comparison = 0

      switch (sortBy) {
        case 'timestamp':
          comparison = new Date(a.created_at).getTime() - new Date(b.created_at).getTime()
          break
        case 'user':
          comparison = (a.user_name || '').localeCompare(b.user_name || '')
          break
        case 'action':
          comparison = a.action_type.localeCompare(b.action_type)
          break
        case 'severity':
          const severityOrder = { 'INFO': 1, 'WARNING': 2, 'ERROR': 3, 'CRITICAL': 4 }
          comparison = (severityOrder[a.severity] || 0) - (severityOrder[b.severity] || 0)
          break
      }

      return sortOrder === 'asc' ? comparison : -comparison
    })

    return sorted
  }

  /**
   * Group audit logs by date
   */
  static groupAuditLogsByDate(auditLogs: AuditLog[]): Record<string, AuditLog[]> {
    const grouped: Record<string, AuditLog[]> = {}
    
    auditLogs.forEach(log => {
      const date = new Date(log.created_at).toDateString()
      if (!grouped[date]) {
        grouped[date] = []
      }
      grouped[date].push(log)
    })
    
    return grouped
  }

  /**
   * Get audit log summary
   */
  static getAuditLogSummary(auditLogs: AuditLog[]): {
    totalLogs: number
    userActions: number
    systemEvents: number
    securityEvents: number
    transactionEvents: number
    criticalEvents: number
    uniqueUsers: number
  } {
    const uniqueUsers = new Set<string>()
    let userActions = 0
    let systemEvents = 0
    let securityEvents = 0
    let transactionEvents = 0
    let criticalEvents = 0

    auditLogs.forEach(log => {
      if (log.user_id) {
        uniqueUsers.add(log.user_id)
        userActions++
      } else {
        systemEvents++
      }

      if (this.isSecurityEvent(log)) {
        securityEvents++
      }

      if (this.isTransactionEvent(log)) {
        transactionEvents++
      }

      if (log.severity === 'CRITICAL') {
        criticalEvents++
      }
    })

    return {
      totalLogs: auditLogs.length,
      userActions,
      systemEvents,
      securityEvents,
      transactionEvents,
      criticalEvents,
      uniqueUsers: uniqueUsers.size
    }
  }

  /**
   * Validate date range for reports
   */
  static validateDateRange(startDate: string, endDate: string): { valid: boolean; error?: string } {
    try {
      const start = new Date(startDate)
      const end = new Date(endDate)

      if (isNaN(start.getTime()) || isNaN(end.getTime())) {
        return { valid: false, error: 'Invalid date format' }
      }

      if (start >= end) {
        return { valid: false, error: 'Start date must be before end date' }
      }

      const daysDiff = Math.ceil((end.getTime() - start.getTime()) / (1000 * 60 * 60 * 24))
      if (daysDiff > 365) {
        return { valid: false, error: 'Date range cannot exceed 365 days' }
      }

      return { valid: true }
    } catch {
      return { valid: false, error: 'Invalid date format' }
    }
  }

  /**
   * Convert audit logs to CSV format
   */
  static convertToCSV(auditLogs: AuditLog[]): string {
    if (auditLogs.length === 0) {
      return ''
    }

    const headers = [
      'Timestamp',
      'User ID',
      'User Name',
      'Action Type',
      'Entity Type',
      'Entity ID',
      'IP Address',
      'Severity',
      'Is System Event'
    ]

    const csvRows = [
      headers.join(','),
      ...auditLogs.map(log => [
        `"${log.created_at}"`,
        `"${log.user_id || ''}"`,
        `"${log.user_name || ''}"`,
        `"${log.action_type}"`,
        `"${log.entity_type}"`,
        `"${log.entity_id || ''}"`,
        `"${log.ip_address || ''}"`,
        `"${log.severity}"`,
        `"${log.is_system_event}"`
      ].join(','))
    ]

    return csvRows.join('\n')
  }

  /**
   * Download data as file
   */
  static downloadAsFile(data: string, filename: string, mimeType: string = 'text/plain'): void {
    const blob = new Blob([data], { type: mimeType })
    const url = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = url
    link.download = filename
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    window.URL.revokeObjectURL(url)
  }
}

export default AuditService